src/libraries/Common/src/System/Security/Cryptography/Oids.Shared.cs

```diff
@@ -67,5 +71,79 @@ private static Oid InitializeOid(string oidValue)
             return oid;
         }
 
+        internal static Oid GetSharedOrNewOid(ref AsnValueReader asnValueReader)
+        {
+            Oid? ret = GetSharedOrNullOid(ref asnValueReader);
+
+            if (ret is not null)
+            {
+                return ret;
+            }
+
+            string oidValue = asnValueReader.ReadObjectIdentifier();
+            return new Oid(oidValue, null);
+        }
+
+        internal static Oid? GetSharedOrNullOid(ref AsnValueReader asnValueReader, Asn1Tag? expectedTag = null)
+        {
+#if NET
+            Asn1Tag tag = asnValueReader.PeekTag();
+
+            // This isn't a valid OID, so return null and let whatever's going to happen happen.
+            if (tag.IsConstructed)
+            {
+                return null;
+            }
+
+            Asn1Tag expected = expectedTag.GetValueOrDefault(Asn1Tag.ObjectIdentifier);
+
+            Debug.Assert(
+                expected.TagClass != TagClass.Universal ||
+                expected.TagValue == (int)UniversalTagNumber.ObjectIdentifier,
+                $"{nameof(GetSharedOrNullOid)} was called with the wrong Universal class tag: {expectedTag}");
+
+            // Not the tag we're expecting, so don't match.
+            if (!tag.HasSameClassAndValue(expected))
+            {
+                return null;
+            }
+
+            ReadOnlySpan<byte> contentBytes = asnValueReader.PeekContentBytes();
+
+            Oid? ret = contentBytes switch
+            {
+                [0x55, 0x04, 0x03] => CommonNameOid,
+                _ => null,
+            };
```

Comment:
This is here as a strawman for improving perf and reducing memory when parsing OIDs.  If we like it, I'll add the rest of the common X.500 attributes in this PR, and we can add more over time.